community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/reader/SimpleIndexReader.java

```diff
@@ -139,6 +141,58 @@ public PrimitiveLongResourceIterator query( IndexQuery... predicates ) throws In
         }
     }
 
+    @Override
+    public IndexProgressor query( NodeValueClient client, IndexQuery... predicates )
+            throws IndexNotApplicableKernelException
+    {
+        IndexQuery predicate = predicates[0];
+        switch ( predicate.type() )
+        {
+        case exact:
+            Value[] values = new Value[predicates.length];
+            for ( int i = 0; i < predicates.length; i++ )
+            {
+                assert predicates[i].type() == exact :
+                        "Exact followed by another query predicate type is not supported at this moment.";
+                values[i] = ((IndexQuery.ExactPredicate) predicates[i]).value();
+            }
+            return seek( client, values );
+        case exists:
+            for ( IndexQuery p : predicates )
+            {
+                if ( p.type() != IndexQueryType.exists )
+                {
+                    throw new IndexNotApplicableKernelException(
+                            "Exists followed by another query predicate type is not supported." );
+                }
+            }
+            return scan( client );
+        case rangeNumeric:
+            assertNotComposite( predicates );
+            IndexQuery.NumberRangePredicate np = (IndexQuery.NumberRangePredicate) predicate;
+            return rangeSeekByNumberInclusive( client, np.from(), np.to() );
+        case rangeString:
+            assertNotComposite( predicates );
+            IndexQuery.StringRangePredicate sp = (IndexQuery.StringRangePredicate) predicate;
+            return rangeSeekByString( client, sp.from(), sp.fromInclusive(), sp.to(), sp.toInclusive() );
+        case stringPrefix:
+            assertNotComposite( predicates );
+            IndexQuery.StringPrefixPredicate spp = (IndexQuery.StringPrefixPredicate) predicate;
+            return rangeSeekByPrefix( client, spp.prefix() );
+        case stringContains:
+            assertNotComposite( predicates );
+            IndexQuery.StringContainsPredicate scp = (IndexQuery.StringContainsPredicate) predicate;
+            return containsString( client, scp.contains() );
+        case stringSuffix:
+            assertNotComposite( predicates );
+            IndexQuery.StringSuffixPredicate ssp = (IndexQuery.StringSuffixPredicate) predicate;
+            return endsWith( client, ssp.suffix() );
+        default:
+            // todo figure out a more specific exception
+            throw new RuntimeException( "Index query not supported: " + Arrays.toString( predicates ) );
```

Comment:
Now would be a good time for that, right?

Reply:
is now the time @burqen @tinwelint ?

Reply:
The time to figure out at more specific exception? Yes, it's always a good time for that! ;)
To be honest though, I can't think of what more to add to that exception.